megablocks/layers/glu.py

```diff
@@ -42,6 +44,139 @@ def forward(self, x, topo):
+        if ctx.num_input_bits == -1:
+            x = saved_tensors[4]
+        else:
+            x_q, x_scales = saved_tensors[4:6]
+
+        # Either 1 or 2 tensors at the end for saved GELU input / sdd output
+        if ctx.num_remat_bits == -1:
+            sdd_out, v1_out = saved_tensors[-2:]
+        else:
+            raise NotImplementedError(f'Activation compression of hidden state not implemented. Set `num_remat_bits = -1`.')
+            # sdd_out_q, sdd_out_scales, v1_out_q, v1_out_scales = saved_tensors[-4:]
+
+        # Rematerialize activation_fn output.
+        activation_fn = ctx.activation_fn
+        activation_grad_fn = None
+        if ctx.num_remat_bits == -1:
+            with torch.set_grad_enabled(True):
+                sdd_out.requires_grad = True
+                v1_out.requires_grad = True
+                def act_glu(x1, x2, activation_fn):
+                    return activation_fn(x1) * x2
+                activation_fn_out = act_glu(sdd_out, v1_out, activation_fn)
+                activation_grad_fn = activation_fn_out.backward
+        else:
+            raise NotImplementedError(f'Activation compression of hidden state not implemented. Set `num_remat_bits = -1`.')
+
+        # Compute dw2 with recomputed activation_fn output.
+        dw2 = gg.backend.gmm(
+            activation_fn_out, ddsd_out, batch_sizes, trans_a=True)
+
+        # Compute dactivation_fn_out.
+        #
+        # NOTE: We reuse the activation_fn_out allocation.
+        dactivation_fn_out = activation_fn_out
+        gg.backend.gmm(
+            ddsd_out, w2, batch_sizes, trans_b=True, c=dactivation_fn_out)
+
+        # Compute dsdd_out.
+        #
+        # NOTE: This reuses the dactivation_fn_out allocation.
+        if ctx.num_remat_bits == -1:
+            assert activation_grad_fn is not None
+            activation_grad_fn(dactivation_fn_out)
+            dsdd_out = sdd_out.grad
+            dv1_out = v1_out.grad
+        else:
+            raise NotImplementedError(f'Activation compression of hidden state not implemented. Set `num_remat_bits = -1`.')
+
+        # rematerialize MLP input now that we need it
+        if ctx.num_input_bits != -1:
+            x = turbo.dequantize_signed(
+                x_q, x_scales, num_bits=ctx.num_input_bits,
+                out_dtype=dtype, out_shape=ctx.x_shape)
+
+        # Compute dw1.
+        dw1 = gg.backend.gmm(dsdd_out, x, batch_sizes, trans_a=True)
+
+        # Compute dv1.
+        dv1 = gg.backend.gmm(dv1_out, x, batch_sizes, trans_a=True)
+
+        # Compute dx.
+        #
+        # NOTE: This reuses the ddsd_out allocation.
+        dx = ddsd_out
+        gg.backend.gmm(dsdd_out, w1, batch_sizes, c=dx)
+        dx += gg.backend.gmm(dv1_out, v1, batch_sizes)
+        return dx, dw1, dv1, dw2, None, None, None, None
+
+memory_optimized_grouped_glu = MemoryOptimizedGroupedGLU.apply
+
+
 class GroupedGLU(SparseGLU):
     def forward(self, x, tokens_per_expert):
```

Comment:
Noticing now - does this function need to scale_grad on the weights?

Reply:
[here](https://github.com/stanford-futuredata/megablocks/pull/66/files#diff-6d3db0eaf79501bae35b15e2735c8cd0225650305971979eae5593c9a300ac30R183) as noted by Sasha